rapid_response_xblock/static/js/src/rapid.js

```diff
@@ -43,6 +46,227 @@
+      colorDomain = [];
+    }
+
+    function makeHistogram(responses) {
+      // Calculate count data for each answer id for all responses.
+      // The returned array is sorted by lowercase answer id.
+      var lookup = {};
+      var uniqueResponses = [];
+      responses.forEach(function(response) {
+        if (!(response.answer_id in lookup)) {
+          lookup[response.answer_id] = 1;
+          uniqueResponses.push(response);
+        } else {
+          lookup[response.answer_id] += 1;
+        }
+      });
+      uniqueResponses = _.sortBy(uniqueResponses, function(response) {
+        return response.answer_id.toLowerCase();
+      });
+
+      return uniqueResponses.map(function(response) {
+        return {
+          answer_id: response.answer_id,
+          answer_text: response.answer_text,
+          count: lookup[response.answer_id]
+        };
+      });
+    }
+
+    function makeIntegerTicks(domain) {
+      // Given the domain limits return 6 or so tick values, equally spaced out, all integers.
+      var tickValues = [];
+      var lookup = {};
+      var maxCount = domain.domain()[1];
+      if (!isNaN(maxCount)) {
+        var numTicks = 6;
+        var tickIncrement = Math.ceil(maxCount / numTicks);
+        for (var tickCount = 0; tickCount < numTicks; ++tickCount) {
+          var tick = tickCount * tickIncrement;
+          if (!(tick in lookup)) {
+            lookup[tick] = true;
+            tickValues.push(tick);
+          }
+        }
+      }
+      return tickValues;
+    }
+
+    function wrapText(textSelector, barWidth, oldText) {
+      // see https://bl.ocks.org/mbostock/7555321 for inspiration
+      // SVG doesn't have a capability to wrap text except for foreignObject which is not supported in IE11.
+      // So we have to calculate it manually
+      textSelector.each(function() {
+        var root = d3.select(this);
+
+        var rootY = root.attr("y");
+        var rootDy = parseFloat(root.attr("dy"));
+
+        root.selectAll("tspan").remove();
+        var words = oldText.split(/\s+/);
+        var tspan = root.append("tspan").attr("x", 0).attr("y", rootY).attr("dy", rootDy + "em");
+
+        var currentLine = 0;
+        var lineHeight = 1.1;
+        words.forEach(function(word) {
+          if (!word) {
+            return;
+          }
+
+          var oldText = tspan.text();
+          tspan.text(oldText + " " + word);
+          if (tspan.node().getComputedTextLength() > barWidth) {
+            currentLine++;
+            tspan.text(oldText + " ");
+            tspan = root.append("tspan").attr("x", 0).attr("y", rootY).attr(
+              "dy", ((currentLine * lineHeight) + rootDy) + "em"
+            ).text(word);
+          }
+        });
+      });
+    }
+
+    function renderD3() {
+      // This function should be called whenever the response information changes.
```

Comment:
1. I think this function should take `responses` as a parameter rather than getting them from the `state` object defined outside of its scope. In fact, we might want to go one further and have `render` take `state` as a parameter, but I'm fine with ignoring that for now
1. This comment should explain what the function does rather than when it should be called. Maybe something like "Renders the graph and adjusts axes based on responses to the given problem."